### ISSUE TITLE
torch.func.vjp: compiled vjp_fn returns incorrect zero gradients

### ISSUE THREAD

**Issue by tristancinquin** [pytorch/pytorch#172386]
### Description

When compiling the function returned by `torch.func.vjp` using torch.compile, the compiled vjp_fn returns incorrect gradients (all zeros).

This is a silent correctness issue:
* eager vjp_fn returns correct gradients
* compiled vjp_fn runs without error
* compiled output is wrong

### Minimal Reproducible Example

```python
import torch
from torch import nn
from torch.func import vjp, functional_call

torch.manual_seed(0)

model = nn.Sequential(
    nn.Linear(4, 1)
)

x = torch.randn(4, 4)
params = {k: v for k, v in model.named_parameters()}

# Functional form
f = lambda p: functional_call(model, p, x)

# Eager VJP works
y, vjp_fn = vjp(f, params)
v = torch.ones_like(y)
gx = vjp_fn(v)[0]
print("eager vjp:", gx) # Non-zero

# Compile the vjp function
vjp_fn_compiled = torch.compile(vjp_fn)
gx2 = vjp_fn_compiled(v)[0]
print("compiled vjp:", gx2)  # all zeros
```

### Observed behavior

* vjp_fn(v) (eager) returns correct non-zero gradients
* torch.compile(vjp_fn)(v) returns a dictionary of all zeros
* No error or warning is raised

### Expected behavior

The compiled vjp_fn should:
* return the same gradients as the eager version
* raise an error

### Error logs

No error is raised

### Versions

PyTorch version: 2.9.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.3.2)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 12:22:05) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-26.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.3
[pip3] torch==2.9.1
[pip3] torch-dct==0.1.6
[pip3] torchaudio==2.9.1
[pip3] torchmetrics==1.8.2
[pip3] torchvision==0.24.1
[conda] numpy                        2.3.3            pypi_0           pypi
[conda] torch                        2.9.1            pypi_0           pypi
[conda] torch-dct                    0.1.6            pypi_0           pypi
[conda] torchaudio                   2.9.1            pypi_0           pypi
[conda] torchmetrics                 1.8.2            pypi_0           pypi
[conda] torchvision                  0.24.1           pypi_0           pypi

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @Chillee @samdow @kshitij12345

**Comment by ayushsatyam146** [comment]
@ezyang @tristancinquin I am working on this. will come up with the fix soon

**Comment by tristancinquin** [comment]
Hi @ayushsatyam146, thank you so much for working on this! I filed the issue but to be honest, I don't have deep enough familiarity with dynamo and functorch internals to meaningfully review the PR. I will defer to @kshitij12345 @ezyang @anijain2305 for the actual review. Appreciate the quick turnaround on the fix!
